src/lib.rs

```diff
@@ -20,49 +52,75 @@ impl Cleverbot {
-                key: key,
-                nick: json.get("nick").unwrap().to_string(),
-            })
-        } else {
-            Err(result.to_string())
+        try!(response.read_to_string(&mut body));
+        let json: BTreeMap<String, String> = try!(serde_json::from_str(&body));
+        let opt_result = json.get("status");
+        let result = match opt_result {
+            Some(result) => result,
+            None => return Err(CleverbotError::MissingValue(String::from("status"))),
+        };
+        match result.as_ref() {
+            "success" => {
+                let json_nick = json.get("nick");
+                let nick = match json_nick {
+                    Some(nick) => nick,
+                    None => return Err(CleverbotError::MissingValue(String::from("nick")))
+                };
+                Ok(Cleverbot {
+                    user: user,
+                    key: key,
+                    nick: nick.to_string(),
+                })
+            },
+            "Error: API credentials incorrect" => Err(CleverbotError::IncorrectCredentials),
+            "Error: reference name already exists" => Err(CleverbotError::DuplicatedReferenceNames),
+            _ => Err(CleverbotError::Api(result.to_owned()))
         }
     }
 
     /// Sends the bot a message and returns its response. If the nick is not set, it will
     /// set it randomly through set_nick_randomly. Returns its response or error string.
     /// * `message` - The message to send to the bot.
-    pub fn say(&mut self, message: &str) -> Result<String, String> {
+    pub fn say(&mut self, message: &str) -> Result<String, CleverbotError> {
         let args = vec![
             ("user", &*self.user),
             ("key", &*self.key),
             ("nick", &*self.nick),
             ("text", message),
         ];
-        let mut response = request("https://cleverbot.io/1.0/ask", &*args);
+        let mut response = {
+            try!(request("https://cleverbot.io/1.0/ask", &*args))
+        };
         let mut body = String::new();
-        response.read_to_string(&mut body).unwrap();
-        let json: BTreeMap<String, String> = serde_json::from_str(&body).unwrap();
-        let result = json.get("status").unwrap();
-        if result == "success" {
-            return Ok(json.get("response").unwrap().to_string());
-        } else {
-            return Err(result.to_string());
+        try!(response.read_to_string(&mut body));
+        let json: BTreeMap<String, String> = try!(serde_json::from_str(&body));
+        let opt_result = json.get("status");
+        let result = match opt_result {
+            Some(result) => result,
+            None => return Err(CleverbotError::MissingValue(String::from("status"))),
+        };
+        match result.as_ref() {
+            "success" => {
+                let json_response = json.get("response");
+                let response = match json_response {
+                    Some(response) => response,
```

Comment:
Similarly here you could just put `Ok(response.to_string())` right away here, letting that `Err` below be an implicit return.